File: ios/RendezvousIL/Services/ChatDataStore.swift
import Foundation

/// On-disk chat cache so channel lists and threads appear instantly, then refresh in the background.
enum ChatDataStore {
    private static var cacheDirectory: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let folder = base.appendingPathComponent("RendezvousIL/Chat", isDirectory: true)
        try? FileManager.default.createDirectory(at: folder, withIntermediateDirectories: true)
        return folder
    }

    private static var channelsURL: URL {
        cacheDirectory.appendingPathComponent("channels.json")
    }

    private static func messagesURL(channelId: String) -> URL {
        let safe = channelId.replacingOccurrences(of: "/", with: "_")
        return cacheDirectory.appendingPathComponent("messages-\(safe).json")
    }

    static func loadChannels() -> [ChatChannelSummary]? {
        guard let data = try? Data(contentsOf: channelsURL) else { return nil }
        return try? JSONDecoder().decode([ChatChannelSummary].self, from: data)
    }

    static func saveChannels(_ channels: [ChatChannelSummary]) {
        guard let data = try? JSONEncoder().encode(channels) else { return }
        try? data.write(to: channelsURL, options: .atomic)
    }

    static func loadMessages(channelId: String) -> [ChatMessage]? {
        guard let data = try? Data(contentsOf: messagesURL(channelId: channelId)) else { return nil }
        return try? JSONDecoder().decode([ChatMessage].self, from: data)
    }

    static func saveMessages(_ messages: [ChatMessage], channelId: String) {
        guard let data = try? JSONEncoder().encode(messages) else { return }
        try? data.write(to: messagesURL(channelId: channelId), options: .atomic)
    }
}

File: android/app/src/main/java/com/rendezvousil/app/chat/ChatDataStore.kt
package com.rendezvousil.app.chat

import android.content.Context
import com.rendezvousil.core.network.dto.ChatChannelSummary
import com.rendezvousil.core.network.dto.ChatMessage
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import java.io.File

/** On-disk chat cache so lists/threads appear instantly, then refresh in the background. */
class ChatDataStore(context: Context) {
    private val dir = File(context.applicationContext.filesDir, "chat-cache").also { it.mkdirs() }
    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    fun loadChannels(): List<ChatChannelSummary>? = readList(channelsFile(), ChatChannelSummary.serializer())

    fun saveChannels(channels: List<ChatChannelSummary>) {
        writeList(channelsFile(), channels, ChatChannelSummary.serializer())
    }

    fun loadMessages(channelId: String): List<ChatMessage>? =
        readList(messagesFile(channelId), ChatMessage.serializer())

    fun saveMessages(channelId: String, messages: List<ChatMessage>) {
        writeList(messagesFile(channelId), messages, ChatMessage.serializer())
    }

    private fun channelsFile() = File(dir, "channels.json")

    private fun messagesFile(channelId: String): File {
        val safe = channelId.replace('/', '_')
        return File(dir, "messages-$safe.json")
    }

    private fun <T> readList(
        file: File,
        serializer: kotlinx.serialization.KSerializer<T>,
    ): List<T>? {
        if (!file.exists()) return null
        return runCatching {
            json.decodeFromString(ListSerializer(serializer), file.readText())
        }.getOrNull()
    }

    private fun <T> writeList(
        file: File,
        items: List<T>,
        serializer: kotlinx.serialization.KSerializer<T>,
    ) {
        runCatching {
            file.writeText(json.encodeToString(ListSerializer(serializer), items))
        }
    }
}
